datafusion/src/physical_plan/expressions/coercion.rs

```diff
@@ -147,6 +147,10 @@ pub fn temporal_coercion(lhs_type: &DataType, rhs_type: &DataType) -> Option<Dat
     }
 }
 
+pub(crate) fn is_dictionary(t: &DataType) -> bool {
```

Comment:
The duplication of code between this module and `binary_rule.rs` is not good -- I am going to try and consolidate it  as a follow on PR cc @liukun4515

Reply:
https://github.com/apache/arrow-datafusion/pull/1607 <-- doesn't remove duplication but it at least consolidates the logic into a single module